src/lib/operators/table_scan/abstract_table_scan_impl.hpp

```diff
@@ -56,6 +63,145 @@ class AbstractTableScanImpl {
+
+    // We assume a maximum SIMD register size of 256 bit. Smaller registers simply lead to the inner loop being unrolled
+    // more than once. Even on machines with 512-bit SIMD registers, we prefer to use 256 bits, because current Intel
+    // CPUs clock down when 512-bit is used.
+    constexpr size_t SIMD_SIZE = 256 / 8;
+    constexpr size_t BLOCK_SIZE = SIMD_SIZE / sizeof(ChunkOffset);
+
+    // The index at which we will write the next matching row
+    auto matches_out_index = matches_out.size();
+
+    // Make sure that we have enough space for the first iteration. We might resize later on.
+    matches_out.resize(matches_out.size() + BLOCK_SIZE, RowID{chunk_id, 0});
+
+    // Continue the following until we have too few rows left to run over a whole block
+    while (static_cast<size_t>(left_end - left_it) > BLOCK_SIZE) {
+      // __mmask8 would be sufficient, but either GCC or the CPU likes a 16-bit mask better, which is good for the
+      // performance.
+      __mmask16 mask = 0;
+
+      // The OpenMP Pragma makes the compiler try harder to vectorize this and gives some hints to help with this.
+      // We do not use the OpenMP runtime, but only the compiler pragmas (look up -fopenmp-simd).
+      // NOLINTNEXTLINE
+      ;  // clang-format off
+      #pragma omp simd reduction(|:mask) safelen(BLOCK_SIZE)
+      // clang-format on
+      for (size_t i = 0; i < BLOCK_SIZE; ++i) {
+        // Fill `mask` with 1s at positions where the condition is fulfilled
+        const auto& left = *left_it;
+
+        if constexpr (std::is_same_v<RightIterator, std::false_type>) {
+          mask |= ((!CheckForNull | !left.is_null()) & func(left)) << i;  // TODO check
+        } else {
+          const auto& right = *right_it;
+          mask |= ((!CheckForNull | (!left.is_null() && !right.is_null())) & func(left, right)) << i;  // TODO check
+        }
+
+        ++left_it;
+        if constexpr (!std::is_same_v<RightIterator, std::false_type>) ++right_it;
+      }
+
+      if (!mask) continue;
+
+      // Next, write *all* offsets in the block into `offsets`
+      auto offsets = std::array<ChunkOffset, BLOCK_SIZE>{};
+
+      if constexpr (!std::is_base_of_v<BasePointAccessSegmentIterator<std::decay_t<decltype(left_it)>,
+                                                                      std::decay_t<decltype(*left_it)>>,
+                                       std::decay_t<decltype(left_it)>>) {
+        // Fast path: If this is a sequential iterator, we know that the chunk offsets are incremented by 1, so we can
+        // save us the memory lookup
+
+        const auto first_offset = (left_it - BLOCK_SIZE)->chunk_offset();
+
+        // NOLINTNEXTLINE
+        ;  // clang-format off
+        #pragma omp simd safelen(BLOCK_SIZE)
+        // clang-format on
+        for (size_t i = 0; i < BLOCK_SIZE; ++i) {
+          offsets[i] = first_offset + static_cast<ChunkOffset>(i);
+        }
```

Comment:
@hendraet @johannes-schneider This is similar to what we want to do with our SortedSegments implementation (i.e., precalculating the offsets instead of looking them up).